### PR TITLE
[CARBONDATA-3833] Make geoID visible

### PR DESCRIPTION
**Pull request by ShreelekhyaG**
### Why is this PR needed?
 To make geohash column visible to the user
 
 ### What changes were proposed in this PR?
Generated geohash column is included in the schema. Validation added to avoid alter and drop the geohash column. Indexes, MV and other table properties not supported on this column.
    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1348/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3069/

**Comment by ydvpankaj99** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3076/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1355/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1362/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3083/

**Comment by ShreelekhyaG** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1366/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3087/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1367/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3088/

**Comment by ShreelekhyaG** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1368/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3089/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3090/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1369/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3104/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1381/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3111/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1387/

File: integration/spark/src/test/scala/org/apache/carbondata/geo/GeoTest.scala [hunk 1]
```diff
@@ -122,7 +161,8 @@ class GeoTest extends QueryTest with BeforeAndAfterAll with BeforeAndAfterEach {
     createTable(sourceTable)
     loadData(sourceTable)
     createTable(targetTable, "'SORT_SCOPE'='GLOBAL_SORT',")
-    sql(s"insert into  $targetTable select * from $sourceTable")
+    sql(s"insert into  $targetTable select timevalue, longitude," +
```
Comment (VenuReddy2103) [inline integration/spark/src/test/scala/org/apache/carbondata/geo/GeoTest.scala:164]: why select * is changed to individual columns ? Probably because you get geospatial column as well when you give select *.  But, we need to handle this insert into select *.

File: integration/spark/src/main/scala/org/apache/carbondata/spark/util/CommonUtil.scala [hunk 2]
```diff
@@ -653,6 +653,21 @@ object CommonUtil {
     storeLocation
   }
 
+  def validateForSpatialTypeColumn(properties: Map[String, String],
```
Comment (VenuReddy2103) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/util/CommonUtil.scala:656]: `columnPropert` string is just to use that in exception message. we alreday have a similar method `validateSpatialIndexColumn()` for it. Suggest to reuse.

File: integration/spark/src/main/scala/org/apache/spark/sql/CarbonSource.scala [hunk 3]
```diff
@@ -379,6 +380,17 @@ object CarbonSource {
     if (isCreatedByCarbonExtension(properties)) {
       // Table is created by SparkSession with CarbonExtension,
       // There is no TableInfo yet, so create it from CatalogTable
+      val columnSchema = updateTable.schema
+      val spatialProperty = properties.get(CarbonCommonConstants.SPATIAL_INDEX)
+      // validate for spatial index column
+      columnSchema.foreach(x => {
+        if (spatialProperty.isDefined &&
```
Comment (VenuReddy2103) [inline integration/spark/src/main/scala/org/apache/spark/sql/CarbonSource.scala:387]: instead of `spatialProperty.isDefined`check inside the loop it should be outside. We can avoid unnecessary loop traversal if the property itself is not defined.

File: integration/spark/src/main/scala/org/apache/spark/sql/CarbonSource.scala [hunk 4]
```diff
@@ -379,6 +380,17 @@ object CarbonSource {
     if (isCreatedByCarbonExtension(properties)) {
       // Table is created by SparkSession with CarbonExtension,
       // There is no TableInfo yet, so create it from CatalogTable
+      val columnSchema = updateTable.schema
```
Comment (VenuReddy2103) [inline integration/spark/src/main/scala/org/apache/spark/sql/CarbonSource.scala:383]: why moved this exception case alone here from `processSpatialIndexProperty` ? Probably we have missed something here. I think, we have many other invalid property exceptions in `processSpatialIndexProperty` and even for other properties in `prepareTableModel()`.

File: integration/spark/src/main/scala/org/apache/spark/sql/catalyst/CarbonParserUtil.scala [hunk 5]
```diff
@@ -213,8 +207,8 @@ object CarbonParserUtil {
 
     // Process spatial index property
     val indexFields = processSpatialIndexProperty(tableProperties, fields)
-    val allFields = fields ++ indexFields
 
+    val allFields = (fields ++ indexFields).distinct
```
Comment (VenuReddy2103) [inline integration/spark/src/main/scala/org/apache/spark/sql/catalyst/CarbonParserUtil.scala:211]: why distinct ? have already checked that indexFields are not in actual fields. right ? If required please add the comments here.

File: integration/spark/src/main/scala/org/apache/spark/sql/catalyst/CarbonParserUtil.scala [hunk 6]
```diff
@@ -337,7 +331,7 @@ object CarbonParserUtil {
 
     if (tableProperties.get(CarbonCommonConstants.COLUMN_META_CACHE).isDefined) {
       // validate the column_meta_cache option
-      val tableColumns = dims.view.filterNot(_.spatialIndex).map(x => x.name.get) ++
+      val tableColumns = dims.view.map(x => x.name.get) ++
```
Comment (VenuReddy2103) [inline integration/spark/src/main/scala/org/apache/spark/sql/catalyst/CarbonParserUtil.scala:334]: if we had not removed this filter, then you wouldn't require modification in `validateColumnMetaCacheFields` or `validateColumnMetaCacheOption` i guess.

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertIntoCommand.scala [hunk 7]
```diff
@@ -170,11 +171,18 @@ case class CarbonInsertIntoCommand(databaseNameOp: Option[String],
       convertedStaticPartition)
     scanResultRdd = sparkSession.sessionState.executePlan(newLogicalPlan).toRdd
     if (logicalPartitionRelation != null) {
-      if (selectedColumnSchema.length != logicalPartitionRelation.output.length) {
+      val properties = table.getTableInfo.getFactTable.getTableProperties.asScala
```
Comment (VenuReddy2103) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertIntoCommand.scala:174]: In `getReArrangedIndexAndSelectedSchema()` above, we make `selectedColumnSchema` and return. In that method, we are excluding SpatialColumn column. That is the reason why the `selectedColumnSchema` & `logicalPartitionRelation.output` do not match here. I think, you need to fix there. then this change wouldn't be required anymore.

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CommonLoadUtils.scala [hunk 8]
```diff
@@ -568,8 +567,7 @@ object CommonLoadUtils {
         .getColumnSchemaList.size()))
         .asInstanceOf[Array[ColumnSchema]]
     } else {
-      colSchema = colSchema.filterNot(x => x.isInvisible || x.isComplexColumn ||
-                                           x.getSchemaOrdinal == -1)
+      colSchema = colSchema.filterNot(x => x.isInvisible || x.isComplexColumn)
```
Comment (VenuReddy2103) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CommonLoadUtils.scala:570]: why remove getSchemaOrdinal check? we do not judge spatial column by schema ordinal as -1. And x.getSchemaOrdinal was not introduced with spatial support PR. Same applies to if case above.

File: integration/spark/src/test/scala/org/apache/carbondata/geo/GeoTest.scala [hunk 9]
```diff
@@ -82,6 +81,45 @@ class GeoTest extends QueryTest with BeforeAndAfterAll with BeforeAndAfterEach {
...
+         | AS 'bloomfilter'
+         | PROPERTIES('BLOOM_SIZE'='640000', 'BLOOM_FPP'='0.00001')
+      """.stripMargin))
+   assert(exception.getMessage.contains(
+      s"Spatial Index column is not supported, column 'mygeohash' is spatial column"))
+  }
+
+  test("test geo table custom properties on geohandler column") {
+    try {
+      createTable(table1,"'COLUMN_META_CACHE' = 'mygeohash',")
```
Comment (VenuReddy2103) [inline integration/spark/src/test/scala/org/apache/carbondata/geo/GeoTest.scala:114]: Test framework supports interction of exceptions. Instead of this try catch and assert, please use intercept like above testcases.

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertIntoCommand.scala [hunk 10]
```diff
@@ -482,6 +490,7 @@ case class CarbonInsertIntoCommand(databaseNameOp: Option[String],
       null
     }
     var createOrderColumns = table.getCreateOrderColumn.asScala
+      .filterNot(_.getColumnSchema.isSpatialColumn)
```
Comment (VenuReddy2103) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertIntoCommand.scala:493]: Please refer above comment. Also you would want to revisit `convertToNoDictionaryToBytes()` and `convertToNoDictionaryToBytesWithoutReArrange()` method in `InputProcessorStepWithNoConverterImpl` class. Without this PR, length of `data` and `dataFields` args of those method were not matching as we not including spatial column. so we had some change in them. With your PR it can be simplified. Please refer PR #3760.

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3197/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1471/

File: integration/spark/src/test/scala/org/apache/carbondata/geo/GeoTest.scala (hunk 1 above)
Comment (ShreelekhyaG) [reply integration/spark/src/test/scala/org/apache/carbondata/geo/GeoTest.scala]: Done

File: integration/spark/src/main/scala/org/apache/carbondata/spark/util/CommonUtil.scala (hunk 2 above)
Comment (ShreelekhyaG) [reply integration/spark/src/main/scala/org/apache/carbondata/spark/util/CommonUtil.scala:662]: removed `validateSpatialIndexColumn()` from AlterTableUtil as validation to add column is no longer required, as spatial column is now part of schema.

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1568/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3305/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3308/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1570/

File: integration/spark/src/main/scala/org/apache/spark/sql/CarbonSource.scala (hunk 3 above)
Comment (ShreelekhyaG) [reply integration/spark/src/main/scala/org/apache/spark/sql/CarbonSource.scala]: removed.

File: integration/spark/src/main/scala/org/apache/spark/sql/catalyst/CarbonParserUtil.scala (hunk 6 above)
Comment (ShreelekhyaG) [reply integration/spark/src/main/scala/org/apache/spark/sql/catalyst/CarbonParserUtil.scala]: yes, made changes.

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertIntoCommand.scala (hunk 7 above)
Comment (ShreelekhyaG) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertIntoCommand.scala:174]: In case of an insert with original schema had to remove to match with the input projection list.

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertIntoCommand.scala (hunk 10 above)
Comment (ShreelekhyaG) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertIntoCommand.scala]: even now `data` and `dataFields` args of the methods don't match when insert with original schema.

**Comment by brijoobopanna** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1717/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3459/

File: integration/spark/src/main/scala/org/apache/spark/sql/catalyst/CarbonParserUtil.scala [hunk 11]
```diff
@@ -677,6 +677,8 @@ object CarbonParserUtil {
         val errorMsg = "range_column not support multiple columns"
         throw new MalformedCarbonCommandException(errorMsg)
       }
+      CommonUtil.validateForSpatialTypeColumn(tableProperties, rangeColumn,
```
Comment (ajantha-bhat) [inline integration/spark/src/main/scala/org/apache/spark/sql/catalyst/CarbonParserUtil.scala:680]: Instead of checking property by property, once all the properties are filled, better to validate at one place  ?

Also I see that for sortcolumns, column_metacache,no_inverted_index it is not handled

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertIntoCommand.scala [hunk 12]
```diff
@@ -170,11 +171,18 @@ case class CarbonInsertIntoCommand(databaseNameOp: Option[String],
       convertedStaticPartition)
     scanResultRdd = sparkSession.sessionState.executePlan(newLogicalPlan).toRdd
     if (logicalPartitionRelation != null) {
-      if (selectedColumnSchema.length != logicalPartitionRelation.output.length) {
+      val properties = table.getTableInfo.getFactTable.getTableProperties.asScala
+      val spatialProperty = properties.get(CarbonCommonConstants.SPATIAL_INDEX)
+      var expectedOutput = logicalPartitionRelation.output
+      if (spatialProperty.isDefined && selectedColumnSchema.size + 1 == expectedOutput.length) {
```
Comment (ajantha-bhat) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertIntoCommand.scala:177]: why the changes in this function ?

As user wanted to created geoSpatial column, we created an extra column, select * from table  should include all the columns.

If the target table don't want geo column, user can specify projections.

**we should not skip spatial column while doing insert into**

@VenuReddy2103 , @ShreelekhyaG

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/CarbonProjectForUpdateCommand.scala [hunk 13]
```diff
@@ -70,12 +71,20 @@ private[sql] case class CarbonProjectForUpdateCommand(
...
+    val properties = carbonTable.getTableInfo.getFactTable.getTableProperties.asScala
+    val spatialProperty = properties.get(CarbonCommonConstants.SPATIAL_INDEX)
     columns.foreach { col =>
       val dataType = carbonTable.getColumnByName(col).getColumnSchema.getDataType
       if (dataType.isComplexType) {
         throw new UnsupportedOperationException("Unsupported operation on Complex data type")
       }
-
+      if (spatialProperty.isDefined) {
+        if (col.contains(spatialProperty.get.trim)) {
```
Comment (ajantha-bhat) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/CarbonProjectForUpdateCommand.scala:82]: why `contains` here ? it's a column name. suppose to be equalsIgnoreCase ?

File: integration/spark/src/main/scala/org/apache/spark/sql/hive/CarbonAnalysisRules.scala [hunk 14]
```diff
@@ -266,16 +266,24 @@ case class CarbonPreInsertionCasts(sparkSession: SparkSession) extends Rule[Logi
       relation: LogicalRelation,
       child: LogicalPlan): LogicalPlan = {
     val carbonDSRelation = relation.relation.asInstanceOf[CarbonDatasourceHadoopRelation]
-    if (carbonDSRelation.carbonRelation.output.size > CarbonCommonConstants
+    val carbonTable = carbonDSRelation.carbonRelation.carbonTable
+    val properties = carbonTable.getTableInfo.getFactTable.getTableProperties.asScala
+    val spatialProperty = properties.get(CarbonCommonConstants.SPATIAL_INDEX)
+    var expectedOutput = carbonDSRelation.carbonRelation.output
+    // have to remove geo column to support insert with original schema
```
Comment (ajantha-bhat) [inline integration/spark/src/main/scala/org/apache/spark/sql/hive/CarbonAnalysisRules.scala:273]: I think no need to remove this column from original schema as it is a visible column , same reason as I mentioned for `insertIntoCommand`

File: integration/spark/src/main/scala/org/apache/spark/sql/hive/CarbonFileMetastore.scala [hunk 15]
```diff
@@ -228,8 +230,15 @@ class CarbonFileMetastore extends CarbonMetaStore {
             c.getClass.getName.equals("org.apache.spark.sql.catalyst.catalog.HiveTableRelation") ||
             c.getClass.getName.equals(
               "org.apache.spark.sql.catalyst.catalog.UnresolvedCatalogRelation")) =>
-        val catalogTable =
+        var catalogTable =
           CarbonReflectionUtils.getFieldOfCatalogTable("tableMeta", c).asInstanceOf[CatalogTable]
+        // remove spatial column from schema
```
Comment (ajantha-bhat) [inline integration/spark/src/main/scala/org/apache/spark/sql/hive/CarbonFileMetastore.scala:235]: same as above, we can register this column as it is visible column

File: integration/spark/src/test/scala/org/apache/carbondata/geo/GeoTest.scala [hunk 16]
```diff
@@ -82,6 +81,33 @@ class GeoTest extends QueryTest with BeforeAndAfterAll with BeforeAndAfterEach {
       case None => assert(false)
     }
   }
+  test("test geo table drop spatial index column") {
+    createTable()
+    loadData()
+    val exception = intercept[MalformedCarbonCommandException](sql(s"alter table $table1 drop columns(mygeohash)"))
+    assert(exception.getMessage.contains(
```
Comment (ajantha-bhat) [inline integration/spark/src/test/scala/org/apache/carbondata/geo/GeoTest.scala:88]: some more test case can be added with mixed case column names for other table property validation added, example range column, bucket column etc

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/command/SICreationCommand.scala [hunk 17]
```diff
@@ -209,6 +209,11 @@ private[sql] case class CarbonCreateSecondaryIndexCommand(
             .get
         }")
       }
+      val isSpatialColPresent = dims.find(x => x.getColumnSchema.isSpatialColumn)
```
Comment (ajantha-bhat) [inline integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/command/SICreationCommand.scala:212]: Please check , I think we need to remove isSpatialColumn from columnSchema. 
we used that mainly for making  it invisible. Now as it is visible. It is just another plan column. Instead we can check if column name is in the table property or not.

@VenuReddy2103 @ShreelekhyaG

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertIntoCommand.scala (hunk 12 above)
Comment (ShreelekhyaG) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertIntoCommand.scala:177]: Yes, select *from table includes all columns including geoSpatial. I have added a testcase for that now. This change is when a user tries to insert with original schema. Like `sql(s"insert into $table1 select 1575428400000,116285807,40084087")`

File: integration/spark/src/test/scala/org/apache/carbondata/geo/GeoTest.scala (hunk 16 above)
Comment (ShreelekhyaG) [reply integration/spark/src/test/scala/org/apache/carbondata/geo/GeoTest.scala:174]: ok done.

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/CarbonProjectForUpdateCommand.scala (hunk 13 above)
Comment (ShreelekhyaG) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/CarbonProjectForUpdateCommand.scala]: changed now.

File: integration/spark/src/main/scala/org/apache/spark/sql/catalyst/CarbonParserUtil.scala (hunk 11 above)
Comment (ShreelekhyaG) [reply integration/spark/src/main/scala/org/apache/spark/sql/catalyst/CarbonParserUtil.scala]: modified the method to validate all table properties in one place as suggested.

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3497/

File: integration/spark/src/test/scala/org/apache/carbondata/geo/GeoTest.scala [hunk 18]
```diff
@@ -112,6 +238,23 @@ class GeoTest extends QueryTest with BeforeAndAfterAll with BeforeAndAfterEach {
       result)
   }
 
+  test("test insert into non-geo table select from geo table") {
```
Comment (ajantha-bhat) [inline integration/spark/src/test/scala/org/apache/carbondata/geo/GeoTest.scala:241]: please add a test case of insert into geo table, where insert rows will not have geo data. but select *  shows geo data

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3514/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1772/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1773/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3515/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3516/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1774/

File: integration/spark/src/test/scala/org/apache/carbondata/geo/GeoTest.scala (hunk 18 above)
Comment (ShreelekhyaG) [reply integration/spark/src/test/scala/org/apache/carbondata/geo/GeoTest.scala:241]: modified existing test case and added validation for the geo column.

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/command/SICreationCommand.scala (hunk 17 above)
Comment (ShreelekhyaG) [reply integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/command/SICreationCommand.scala]: removed isSpatialColumn from schema.

File: integration/spark/src/main/scala/org/apache/spark/sql/hive/CarbonFileMetastore.scala (hunk 15 above)
Comment (ShreelekhyaG) [reply integration/spark/src/main/scala/org/apache/spark/sql/hive/CarbonFileMetastore.scala:235]: Here, catalogTable will have spatial column in schema which is used to build carbon table.  As spatial column is not supposed to be present in user-defined columns, removing it here. Later from tableproperties the column will be added in carbonTable.

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/command/SICreationCommand.scala [hunk 19]
```diff
@@ -210,6 +210,14 @@ private[sql] case class CarbonCreateSecondaryIndexCommand(
             .get
         }")
       }
+      val properties = carbonTable.getTableInfo.getFactTable.getTableProperties.asScala
+      val spatialProperty = properties.get(CarbonCommonConstants.SPATIAL_INDEX)
+      if (spatialProperty.isDefined) {
+        if (dims.find(x => x.getColName.equalsIgnoreCase(spatialProperty.get.trim)).isDefined) {
```
Comment (VenuReddy2103) [inline integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/command/SICreationCommand.scala:216]: you would want to check for index column names.

Comment (ShreelekhyaG) [reply integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/command/SICreationCommand.scala]: changed now

File: integration/spark/src/main/scala/org/apache/spark/sql/hive/CarbonAnalysisRules.scala (hunk 14 above)
Comment (ShreelekhyaG) [reply integration/spark/src/main/scala/org/apache/spark/sql/hive/CarbonAnalysisRules.scala:273]: same reason as above. To support inset without geo column.

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3537/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1797/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3544/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1804/

**Comment by ShreelekhyaG** [comment]
retest this please

File: integration/spark/src/main/scala/org/apache/spark/sql/hive/CarbonFileMetastore.scala (hunk 15 above)
Comment (ajantha-bhat) [reply integration/spark/src/main/scala/org/apache/spark/sql/hive/CarbonFileMetastore.scala:235]: would have added this as comment !

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3554/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1814/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3561/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1822/

**Comment by ShreelekhyaG** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1827/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3566/

**Comment by VenuReddy2103** [comment]
LGTM

**Comment by ajantha-bhat** [comment]
LGTM

This PR can break backward compatibility in case of upgrade if anyone is using the geo spatial feature. But As I know it is an alpha feature. so breaking compatibility is OK.

**Comment by ajantha-bhat** [comment]
waiting for build to pass for the merge

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3581/

**Comment by ShreelekhyaG** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3583/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1844/

**Comment by ajantha-bhat** [comment]
Merging this. Thanks for your contribution!
